### ISSUE TITLE
Multiple threads (each with a uring) & `crossbeam::deque` & a pipeline of I/O steps

### ISSUE THREAD

**Issue by JackKelly** [JackKelly/light-speed-io#61]
## Use-cases
- For `get`, we need to get the filesize for a batch of operations using io_uring (#41) whilst concurrently loading other operations. (We'd maybe want to add a `filesize_bytes: Option<u64>` field to `Operation::Get`). 
- When we're loading many non-contiguous small chunks from a file, we'll want to `open` the file in one step, and read in another step. Possibly using separate threads for each step: one uring per thread.

## Implementation and questions

- Users submit tasks using a [`crossbeam::deque::Injector`](https://docs.rs/crossbeam/latest/crossbeam/deque/index.html) (or `crossbeam::deque` or a `Stream`?)
- On startup, we'd create several worker threads, each with its own io_uring? (but how do we also have multiple threads per processing step? ~~Maybe each processing step is actually a Rayon threadpool, where each thread in that threadpool has its own io_uring, because we can't share io_urings across threads?~~)
- ~~Tasks first arrive in a "orchestrator" thread? Which, somehow, wires up other threads to do the work? (e.g. a simple "get filesize" operation would just go to one threadpool. But an "open file and read lots of non-contiguous chunks" operation would go to three steps: 1) get filesize, 2) open (steps 1 and 2 can happen concurrently), 3) read-and-close (chained in io_uring).~~
- UPDATE: Don't use Rayon. Just have a user-settable number of worker threads, each with its own io_uring. Threads pop `Operation`s off the `Injector` FIFO. For multi-step operations (like get filesize, open, read lots of chunks), use the same thread. Each thread will also have an local queue of tasks to submit to its io_uring "ASAP". These tasks will always take precedence over the new tasks arriving from the `crossbeam::channel`. UPDATE: This sounds like what [`crossbeam::deque`](https://docs.rs/crossbeam/latest/crossbeam/deque/index.html) was designed for, maybe I should use that. So, maybe the thread would look something like this:
    - The thread pops a `Task` from the `Injector` FIFO queue.
    - If this `Task` is a `get` then the thread checks if its uring can accommodate any more files. ~~If not, put the `Task` onto the `Stealer` queue (shared with other threads. Or, if possible, return it to the front of the `Injector` queue).~~ UPDATE: Actually, we can simplify this. Just program the thread not to pop any more tasks off the Injector if the thread can't handle any more files in flight. Then we don't need to "re-queue" tasks. 
    - Assuming this is a `get` operation...
    - The thread first submits a "get filesize" op to its uring. And, if other `Tasks` are waiting in the `Injector` queue, then submit those tasks to the uring too, as long as we don't exceed the number of files this uring can handle.
    - When the "get filesize" CQE emerges, allocate a buffer (or re-use a buffer and/or use a registered buffer), and then  add all the `read` ops to the thread's local task queue.
    - When the final `read` op completes, then submit a `close` op.
    - While this is happening, this thread won't take tasks off the user queue. So other threads can take the load.

## Cross references
- #32
- #33
- #44
- #60

**Comment by JackKelly** [comment]
What happens if we want to read 1 million chunks from a single file. As currently designed, that will all be handled by a single thread, and other threads will sit idle.

Some questions to answer:

- is it possible for multiple urings to read in parallel from a single file?
    - I think we'd open the file once, without using a fixed descriptor. Ie, one thread will open the file, and then share that "normal" file descriptor with other worker threads. 
    - each worker thread can then use register_files to register that "normal" FD. 
- does it improve performance to use multiple threads to read a single file?
    - I should do some fio experiments. Although I'm not sure if fio can read from a single file from multiple threads?
- I had originally thought that there would be a 1:1 relationship between tasks and files. So reading 1 millions chunks from a single file would be represented by a single task (in the Injector FIFO). But then we have to have a Stealer queue for threads to share work. Maybe it's easier if we just have a single Task be a single read operation (ie a single contiguous range of bytes). And the thread will need to check if the filename for each new task is a file that the thread already has open. Or maybe it's simpler to go back to the idea of having an "open" step?

**Comment by JackKelly** [comment]
On the injector queue, there will be on file per Operation. Reading a million chunks from one file will be represented by one Operation. 

The thread that receives that Operation will break it into many Operations: On the Worker and Stealer queues, each op will allow multiple `Operation`s per file. Each `Operation` will represent a single `read` or `readv` opcode. These smaller ops will probably stay on that thread. They'll only be stolen if other threads are idle. So it naturally solves the dilemma that, if the workload is to open many files, then it's most efficient to only open one file per thread. But if the workload is to open many chunks from one file, then we do want to spread that across threads. 

I think I've figured out how to know when to close each file: Each `Operation` will have a u32 to say "this operation is one of n ops on this file". When a thread completes an op, it will send a message to the "closer" thread to say "I've completed 1 op out of n ops on this filename". When the closer thread sees n of these messages, then it'll broadcast a message to all threads to say "close this filename". If a thread hasn't opened this filename then that thread just ignores the message.

To open the file: each thread will keep an internal Map, that maps from filename to fixed file handle. If the filename doesn't exist in the map then the thread first submits an open op to uring, and then uses a strictly internal (no steeling) thread to read the file.
